[App] accept option '--' to mark all following arguments as files (refs #3714)

SebDieBln opened the pull request:
If the user wanted to load a file that is named like an option, this was not possible, because the command line argument was always interpreted as an option. This PR makes it possible.

As before, every command line argument is checked whether it matches one of the known options (e.g. `--help`) and if it doesn't, it is treated as a file to load, either project or layer.
Additionally the new option `--` is introduced and every argument following it is treated as a filename.
So to load a layer named _--help_ the correct arguments are now `-- --help`.

This PR includes a test case for the new functionality and some minor improvements to the test.

SebDieBln committed:
improved TestPyQgsAppStartup
allow passing additional arguments after testfile
do not wait the full timeout when the app already exited

SebDieBln committed:
introduce a new command line option '--' to the main app
all arguments after this option are treated as files, not options
includes a test

SebDieBln commented:
Note: This is somewhat related to PR #2491.

nyalldawson commented:
Can you rebase against master? Travis was failing through no fault of this PR

nyalldawson commented:
BTW - just wanted to say thanks for the great stuff you've been submitting lately. It's much appreciated!

SebDieBln commented:
> @nyalldawson commented 
> 
> Can you rebase against master? Travis was failing through no fault of this PR

I rebased and Travis is happy.

I was kind of confused: Travis showed the green light for the branch in my fork, but not for the PR. Then I found out that PRs are merged into master first and then passed to Travis. My fork must have been behind by just the right number of commits :smile:

> BTW - just wanted to say thanks for the great stuff you've been submitting lately. It's much appreciated!

It's my pleasure! QGIS is such a great project to work with and now also to contribute to. And this is because of people like you and all the others. I somehow regret not having been to Gran Canaria. Well, next time for sure!

SebDieBln commented:
@wonder-sk With only a week left until feature freeze, is there anything I can do to help you review and merge this PR?

DelazJ commented:
Indeed! Thanks for the work you all are doing.
Your reactivity is much appreciated

BTW, shouldn't @SebDieBln be listed among Redmine developers (so that some issues can be assigned to him :) )?

SebDieBln commented:
@DelazJ :yum:
BTW: I already have some permissions to modify tickets and I am in the list of assignable people on Redmine. See [Redmine #13887](https://hub.qgis.org/issues/13887) where I changed some attributes - and reverted them again.

Not sure if these are default permissions or if my account got messed up by something.

DelazJ commented:
Weird, I didn't find you earlier and be sure that I looked for you (reported some symbols issues :) )

About Redmine, I also wonder if there is a bug because since yesterday, I have rights to modify properties of any report. Before, I could just open/close and maybe change the version (sorry I don't remember), but clearly few properties even if it was my issue report (and though I was already among possible assignees).
Does this change have something to do with the fact that I join QGIS group for Documentation and website github repo the day before? It shouldn't be that automatic. Or is it Redmine bug?

SebDieBln commented:
Although I tagged it _FEATURE_ this PR partly fixes the _BUG_ [Redmine #3714](https://hub.qgis.org/issues/3714). Could it still be merged for 2.14?
